README.md
# JsMaps - Interactive Javascript Maps
[![npm version](https://img.shields.io/npm/v/jsmaps.svg)](https://www.npmjs.com/package/jsmaps)
![Bower version](https://img.shields.io/bower/v/jsmaps.svg)
[![NuGet](https://img.shields.io/nuget/v/jsmaps.svg)](https://www.nuget.org/packages/jsmaps)
[![License](https://img.shields.io/npm/l/jsmaps.svg)](https://raw.githubusercontent.com/markkr125/jsmaps/master/LICENSE)

JsMaps is a Javascript library with a full featured Interactive Slippy Map, with no external dependencies, only needs a tile server url. JsMaps also provides a single, common interface for a wide variety of Javascript map APIs, allowing a developer to easliy switch from one maps API to another.

[Project Home Page (includes examples)](http://jsmaps.net) 

[API Reference](https://github.com/markkr125/jsmaps/wiki/API-Reference)

## Installation

Several installation options are available:

* [Download the latest release](https://github.com/markkr125/jsmaps/releases).
* Clone the repo: `git clone https://github.com/markkr125/jsmaps.git`.
* Install with [Bower](http://bower.io): `bower install jsmaps`.
* Install with [npm](https://www.npmjs.com): `npm install jsmaps`.
* Install with [NuGet](https://www.nuget.org): `Install-Package jsmaps`.
* Install with [Meteor](https://www.meteor.com): `meteor add mark125:jsmaps`.

## Supported External APIs
* [Google v3](https://developers.google.com/maps/)
* [Here Maps v3](https://developer.here.com/) 
* [Bing v7](https://msdn.microsoft.com/en-us/library/dd877180.aspx)
* [Yandex Maps v2.1](https://tech.yandex.com/maps/)

## Browsers
Browser support depends on the mapping api used. See bellow for documentation regarding the browser support of every API.
* Native Backed - All modern browesrs, and ie8.
* [Google v3](https://developers.google.com/maps/faq#browsersupport)
* [Here Maps v3](https://developer.here.com/javascript-apis/documentation/v3/maps/topics/overview.html) 
* [Bing v7](https://msdn.microsoft.com/library/gg427618.aspx)
* [Yandex Maps v2.1](https://tech.yandex.com/maps/doc/jsapi/2.1/versions/concepts/index-docpage/#browsers)

## License
Licensed under the [The MIT License (MIT)](http://opensource.org/licenses/MIT)


CHANGELOG.md
# Changelog

## 0.0.15 - 2016-10-06
 * native map: stop moving the map on mouseup outside the map (with ie11 fix)
 * native map: get rid of the ugly cursor hack, ie will have the cursor from a custom url
 * fixing issue #1
 * New change options function for the map object
 * make sure dragging vectors on bing and here maps stops on mouseup
 * new feature: adding a "setStyle" for all vectors on all backends.
 * all vectors on all backends have a "getStyle" function
 * Fix for issue #2
 
## 0.0.14 - 2016-02-17
 - Version bump due to meteor package generation
 
 changes:
 * native map: basic pointer events support, the map should work fine on ie10+
 * native map: adding grab and grabbing cursors on the map for dragging effects
 * native map: when the left button is not clicked dont drag the map.
 * native map: remove editable markers when a vector is moving, and put then back only when dragging is done, this solves slow dragging on old ie (bugfix)
 * native map: editable vector markers should respond on the first touch
 * native map: vectors should not move around on map resize with svg backend
 * native map: vectors should be properly draggable after orientationchange event 
 * native map: Mouse scroll should work fine with a small map 
 * native map: if the zoom ui is to big for the map viewport, the zoom handle will be hidden and only +/- buttons will be displayed
 * native map: edit marker on vectors objects will resize the vector even if the marker is dragged outside of the map bounds
 
## 0.0.11 - 2016-02-05
 * native map: circle is editable 
 * native map: many bug fixes on touch based systems
 * native map: zoom and layers ui now work touch based systems
 * native map: vectors are editable and draggable on touch based systems
 * native map: adding an opacity transition for images on movement
 * native map: map click events to touch events
 * bing map: fixing a bug on editable polygon under old ie
 * general api: editable marker that are used on vectors looks a bit better now
 * bing and here maps: circles are editable and draggable

## 0.0.10 - 2016-01-19
 * Only to update the package on nuget

## 0.0.9 - 2016-01-19
Native map changes:
 * will always check for css 3d
 * if no css 3d then check for css 2d
 * if no css 2d, then just show/hide the layer div 
 * using css transition for fadeout effect, javascript fadeout is still used when css transition is not supported
 * use a proper svg path to generate a circle.
 * vml backend now makes a real circle


jsfiles/exampleLoader.js
function formatXml(xml) {
    var formatted = '';
    var reg = /(>)(<)(\/*)/g;
    xml = xml.replace(reg, '$1\r\n$2$3');
    var pad = 0;
    jQuery.each(xml.split('\r\n'), function(index, node) {
        var indent = 0;
        if (node.match( /.+<\/\w[^>]*>$/ )) {
            indent = 0;
        } else if (node.match( /^<\/\w/ )) {
            if (pad != 0) {
                pad -= 1;
            }
        } else if (node.match( /^<\w[^>]*[^\/]>.*$/ )) {
            indent = 1;
        } else {
            indent = 0;
        }

        var padding = '';
        for (var i = 0; i < pad; i++) {
            padding += '  ';
        }

        formatted += padding + node + '\r\n';
        pad += indent;
    });

    return formatted;
}

function htmlEntities(str) {
    return String(str).replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/"/g, '&quot;');
}

var ExampleLoader = {
    jsFiles: function (supplier) {
        var files = {
            'native': ['<link type="text/css" rel="stylesheet" href="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/native/core.native.css" />',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/core.abstract.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/core.abstract-helper.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/core.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/native/core.native.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/native/core.native-ui.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/native/core.native-helper.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/native/core.native-marker.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/native/core.native-infoWindow.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/native/core.native-vector.js" type="text/javascript"></script>'
            ],
            'google': [
                '<script src="http://maps.google.com/maps/api/js?sensor=false" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/core.abstract.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/core.abstract-helper.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/core.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/google/core.google.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/google/core.google-geocoder.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/google/core.google-staticmap.js" type="text/javascript"></script>'
            ],
            'yandex': [
                '<script src="http://api-maps.yandex.ru/2.1/?lang=en_US" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/core.abstract.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/core.abstract-helper.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/core.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/yandex/core.yandex.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/yandex/core.yandex-geocoder.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/yandex/core.yandex-staticmap.js" type="text/javascript"></script>'
            ],
            'bing': [
                '<script type="text/javascript" src="http://ecn.dev.virtualearth.net/mapcontrol/mapcontrol.ashx?v=7.0"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/core.abstract.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/core.abstract-helper.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/core.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/bing/core.bing.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/bing/core.bing-geocoder.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/bing/core.bing-staticmap.js" type="text/javascript"></script>'
            ],
            'here': [
                '<link rel="stylesheet" type="text/css" href="https://js.api.here.com/v3/3.0/mapsjs-ui.css" />',
                '<script type="text/javascript" charset="UTF-8" src="https://js.api.here.com/v3/3.0/mapsjs-core.js"></script>',
                '<script type="text/javascript" charset="UTF-8" src="https://js.api.here.com/v3/3.0/mapsjs-service.js"></script>',
                '<script type="text/javascript" charset="UTF-8" src="https://js.api.here.com/v3/3.0/mapsjs-ui.js"></script>',
                '<script type="text/javascript" charset="UTF-8" src="https://js.api.here.com/v3/3.0/mapsjs-mapevents.js"></script>',
                '<script src="http://js.api.here.com/v3/3.0/mapsjs-pano.js" type="text/javascript" charset="utf-8"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/core.abstract.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/core.abstract-helper.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/core.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/here/core.here.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/here/core.here-geocoder.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/here/core.here-staticmap.js" type="text/javascript"></script>',
                '<script src="https://cdn.rawgit.com/markkr125/jsmaps/'+LIBRARY_VERSION+'/library/here/core.here-vector.js" type="text/javascript"></script>'
            ]
        };

        if (typeof files[supplier] == 'undefined') throw 'unknown supplier '+supplier;
        return files[supplier];
    },

    exampleLoad: function (exampleCode,supplier) {
        exampleCode = decodeURIComponent(exampleCode);
        var staticFiles = this.jsFiles(supplier);

        var html = '<?xml version="1.0" encoding="UTF-8"?>'+
            '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">'+
            '<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en"><head><title>Example</title><style>html, body, #map {width:100%;height: 100%; margin: 0; padding: 0;}</style>';

        for (var i in staticFiles) {
            if (staticFiles.hasOwnProperty(i) == false) continue;
            html += staticFiles[i];
        }


        var toReplace = '';
        var toReplace2 = '{}';

        if (supplier == 'native') {
            toReplace = 'var tiles = new jsMaps.Native.Tiles();'+"\n"+
            'tiles.addTileLayer("http://{s}.tile.osm.org/{z}/{x}/{y}.png",[\'a\',\'b\',\'c\'],\'&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap contributors</a>.\',\'OpenStreetMap\');'+"\n"+
            'tiles.addTileLayer("http://{s}.mqcdn.com/tiles/1.0.0/map/{z}/{x}/{y}.png",[\'otile1\',\'otile2\',\'otile3\',\'otile4\'],\'&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap contributors</a>. Tiles courtesy of <a href="http://www.mapquest.com/" target="_blank">MapQuest</a> <img src="https://developer.mapquest.com/content/osm/mq_logo.png">\',\'Map Quest\');'+"\n"+
            'tiles.addTileLayer("http://{s}.mqcdn.com/tiles/1.0.0/sat/{z}/{x}/{y}.jpg",[\'oatile1\',\'oatile2\',\'oatile3\',\'oatile4\'],\'&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap contributors</a>. Tiles courtesy of <a href="http://www.mapquest.com/" target="_blank">MapQuest</a> <img src="https://developer.mapquest.com/content/osm/mq_logo.png">\',\'Map Quest Satellite\',19);'+"\n";

            toReplace2 = 'tiles';
        } else if (supplier == 'bing') {
            toReplace = 'jsMaps.config.bing.key = \'your-bing-key\';';
        } else if (supplier == 'here') {
            toReplace = 'jsMaps.config.here.app_id = \'here-app-id\';'+"\n"+
            'jsMaps.config.here.app_code = \'here-app-code\';'+"\n";
        }

        html += '</head><body>'+exampleCode+'</body></html>';

        html = html.replace('// PlaceHolder',toReplace);
        html = html.replace('%supplier%',supplier);
        html = html.replace("{'placeholder':1}",toReplace2);

        var writeHtml = html;
        writeHtml = writeHtml.replace('your-bing-key','AvZb3DV_bZiBacG2jm6QNHnz58WwT6X5KfCRKd0a-dnDqihgkEXaowWTasMyN6Io');
        writeHtml = writeHtml.replace('here-app-id','ycU7pzqrEYaipnlczp6c');
        writeHtml = writeHtml.replace('here-app-code','PC2xll3XVkKXM0zX324KMQ');

        //var doc = document.getElementById('example-frame').contentWindow.document;
        //doc.open();
        //doc.write(writeHtml);
        //doc.close();
        var iFrame = document.getElementById('example-frame');
        iFrame.contentWindow.contents = writeHtml;
        iFrame.src = 'javascript:window["contents"]';

        var $fullHtml = $('#fullHtml');

        $fullHtml.find('pre code').text(formatXml(html));

        var re = /<script\b[^>]*>([\s\S]*?)<\/script>/gm;

        var match;
        while (match = re.exec(html)) {
            if (match[1]!='') {
                $('#Javascript').find('code').html(htmlEntities($.trim(match[1])));
            }
        }

        $('pre code').each(function (i, block) {
            hljs.highlightBlock(block);
        });
    },

    exampleButtons: function (exampleCode,allowedSuppliers) {
        var buttons = '';
        var suppliers = ['native','google', 'yandex', 'bing', 'here'];

        if (typeof allowedSuppliers != 'undefined' && allowedSuppliers != 'ALL') {
            suppliers = allowedSuppliers.split(' ');
        }

        for (var x in suppliers) {
            if (suppliers.hasOwnProperty(x) == false) continue;
            var active = (x == 0) ? ' active': '';
            buttons += '<button type="button" class="btn btn-primary'+active+'" data-provider="'+suppliers[x]+'" aria-pressed="false" autocomplete="off">'+suppliers[x]+'</button>';
        }

        var btnGrp = $('#suppliers-btn-group');
        btnGrp.html(buttons);

        btnGrp.find('button').click(function () {
            btnGrp.find('button.active').removeClass('active');
            $(this).addClass('active');

            var supplier = $(this).data('provider');
            ExampleLoader.exampleLoad(exampleCode, supplier);
        });

        btnGrp.find('button').first().click();
    }
};